File: app/src/main/java/org/autojs/autojs/timing/TimedTaskScheduler.java
package org.autojs.autojs.timing;

import android.content.Context;
import android.content.Intent;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.work.Data;
import androidx.work.ExistingWorkPolicy;
import androidx.work.OneTimeWorkRequest;
import androidx.work.WorkManager;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

import org.autojs.autojs.external.ScriptIntents;

import java.util.concurrent.TimeUnit;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

public class TimedTaskScheduler {

    private static final String LOG_TAG = "TimedTaskScheduler";
    private static final long SCHEDULE_TASK_MIN_TIME = TimeUnit.DAYS.toMillis(2);
    private static final String WORK_DATA_TASK_ID = "task_id";

    public static void checkTasks(Context context, boolean force) {
        Log.d(LOG_TAG, "check tasks: force = " + force);
        TimedTaskManager.getInstance().getAllTasks()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(timedTask -> scheduleTaskIfNeeded(context, timedTask, force));
    }

    public static void scheduleTaskIfNeeded(Context context, TimedTask timedTask, boolean force) {
        long millis = timedTask.getNextTime();
        if ((!force && timedTask.isScheduled()) || millis - System.currentTimeMillis() > SCHEDULE_TASK_MIN_TIME) {
            return;
        }
        scheduleTask(context, timedTask, millis, force);
        TimedTaskManager.getInstance()
                .notifyTaskScheduled(timedTask);
    }

    private synchronized static void scheduleTask(Context context, TimedTask timedTask, long millis, boolean force) {
        if (!force && timedTask.isScheduled()) {
            return;
        }
        long timeWindow = millis - System.currentTimeMillis();
        timedTask.setScheduled(true);
        TimedTaskManager.getInstance().updateTaskWithoutReScheduling(timedTask);
        
        if (timeWindow <= 0) {
            runTask(context, timedTask);
            return;
        }
        
        cancel(timedTask);
        Log.d(LOG_TAG, "schedule task: task = " + timedTask + ", millis = " + millis + ", timeWindow = " + timeWindow);

        Data inputData = new Data.Builder()
                .putLong(WORK_DATA_TASK_ID, timedTask.getId())
                .build();

        OneTimeWorkRequest workRequest = new OneTimeWorkRequest.Builder(TimedTaskWorker.class)
                .setInitialDelay(timeWindow, TimeUnit.MILLISECONDS)
                .setInputData(inputData)
                .addTag(String.valueOf(timedTask.getId()))
                .build();

        WorkManager.getInstance(context)
                .beginUniqueWork(String.valueOf(timedTask.getId()), ExistingWorkPolicy.REPLACE, workRequest)
                .enqueue();
    }

    public static void cancel(TimedTask timedTask) {
        WorkManager.getInstance().cancelAllWorkByTag(String.valueOf(timedTask.getId()));
        Log.d(LOG_TAG, "cancel task: task = " + timedTask);
    }

    public static void init(Context context) {
        checkTasks(context, true);
    }

    private static void runTask(Context context, TimedTask task) {
        Log.d(LOG_TAG, "run task: task = " + task);
        Intent intent = task.createIntent();
        ScriptIntents.handleIntent(context, intent);
        TimedTaskManager.getInstance().notifyTaskFinished(task.getId());
    }

    public static class TimedTaskWorker extends Worker {
        private final Context mContext;

        public TimedTaskWorker(@NonNull Context context, @NonNull WorkerParameters params) {
            super(context, params);
            mContext = context;
        }

        @NonNull
        @Override
        public Result doWork() {
            long id = getInputData().getLong(WORK_DATA_TASK_ID, -1);
            TimedTask task = TimedTaskManager.getInstance().getTimedTask(id);
            Log.d(LOG_TAG, "doWork: id = " + id + ", task = " + task);
            if (task == null) {
                return Result.failure();
            }
            runTask(mContext, task);
            return Result.success();
        }
    }
}

File: app/src/main/java/org/github/autojs/timing/TimedTaskScheduler.kt
package org.github.autojs.timing

import android.annotation.SuppressLint
import android.app.AlarmManager
import android.app.PendingIntent
import android.content.Context
import android.os.Build
import android.os.SystemClock
import com.stardust.app.GlobalAppContext
import org.autojs.autojs.external.ScriptIntents
import org.autojs.autojs.timing.TimedTaskManager
import timber.log.Timber
import java.util.concurrent.TimeUnit
import androidx.work.ExistingPeriodicWorkPolicy
import androidx.work.PeriodicWorkRequestBuilder
import androidx.work.WorkManager

object TimedTaskScheduler {

    private const val SCHEDULE_TASK_MIN_TIME = 172_800_000L
    private const val WORK_NAME_CHECK_TASKS = "checkTasks"

    @SuppressLint("CheckResult")
    fun checkTasks(context: Context, force: Boolean) {
        Timber.d("check tasks: force = $force")
        TimedTaskManager.getInstance().allTasks
            .subscribeOn(io.reactivex.schedulers.Schedulers.io())
            .observeOn(io.reactivex.android.schedulers.AndroidSchedulers.mainThread())
            .subscribe { timedTask -> scheduleTaskIfNeeded(context, timedTask, force) }
    }

    @JvmStatic
    fun scheduleTaskIfNeeded(context: Context, timedTask: TimedTask, force: Boolean) {
        val millis = timedTask.getNextTime()
        if ((!force && timedTask.isScheduled) || millis - System.currentTimeMillis() > SCHEDULE_TASK_MIN_TIME) {
            return
        }
        scheduleTask(context, timedTask, millis, force)
        TimedTaskManager.getInstance().notifyTaskScheduled(timedTask)
    }

    @Synchronized
    private fun scheduleTask(context: Context, timedTask: TimedTask, millis: Long, force: Boolean) {
        if (!force && timedTask.isScheduled) return
        val timeWindow = millis - System.currentTimeMillis()
        timedTask.isScheduled = true
        TimedTaskManager.getInstance().updateTaskWithoutReScheduling(timedTask)
        if (timeWindow <= 0) {
            runTask(context, timedTask)
            return
        }
        cancel(timedTask)
        Timber.d("schedule task: task = $timedTask, millis = $millis, timeWindow = $timeWindow")
        val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        val pendingIntent = timedTask.createPendingIntent(context)
        val triggerAtMillis = SystemClock.elapsedRealtime() + timeWindow
        scheduleAlarm(alarmManager, pendingIntent, triggerAtMillis)
    }

    private fun scheduleAlarm(alarmManager: AlarmManager, pendingIntent: PendingIntent, triggerAtMillis: Long) {
        val type = AlarmManager.ELAPSED_REALTIME_WAKEUP
        when {
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.S && alarmManager.canScheduleExactAlarms() ->
                alarmManager.setExactAndAllowWhileIdle(type, triggerAtMillis, pendingIntent)
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.S ->
                alarmManager.setAndAllowWhileIdle(type, triggerAtMillis, pendingIntent)
            else ->
                alarmManager.setExactAndAllowWhileIdle(type, triggerAtMillis, pendingIntent)
        }
    }

    @JvmStatic
    fun cancel(timedTask: TimedTask) {
        val context = GlobalAppContext.get()
        val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        val pendingIntent = timedTask.createPendingIntent(context)
        alarmManager.cancel(pendingIntent)
        Timber.d("cancel task: task = $timedTask")
    }

    fun init(context: Context) {
        val checkWork = PeriodicWorkRequestBuilder<CheckTasksWorker>(
            20, TimeUnit.MINUTES
        ).build()
        WorkManager.getInstance(context)
            .enqueueUniquePeriodicWork(WORK_NAME_CHECK_TASKS, ExistingPeriodicWorkPolicy.KEEP, checkWork)
        checkTasks(context, true)
    }

    private fun runTask(context: Context, task: TimedTask) {
        Timber.d("run task: task = $task")
        val intent = task.createIntent(context)
        ScriptIntents.handleIntent(context, intent)
        TimedTaskManager.getInstance().notifyTaskFinished(task.id)
    }
}
